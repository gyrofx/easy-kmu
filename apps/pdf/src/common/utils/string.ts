import { remove } from 'remove-accents'

/**
 * Trims `text` to a `maxLength`, but keeps the first and last couple of
 * characters. The `maxLength` must be greater than 200 to have an effect
 */
export function summarizeStringTo(text: string, maxLength: number) {
  if (maxLength < 200) throw new Error(`maxLength must be at least 200, got ${maxLength}`)

  if (typeof text !== 'string' || text.length <= maxLength) return text

  const cushionLength = 100
  const startAndEndLengths = Math.floor((maxLength - cushionLength) / 2)
  const start = text.slice(0, startAndEndLengths)
  const end = text.slice(-startAndEndLengths)
  return `[Summarized string of length ${text.length}]: ${start} [...] ${end}`
}

/**
 * @examples
 * removeAccentsAndSpecialCharacters('abc') // 'abc'
 * removeAccentsAndSpecialCharacters('a b c') // 'a-b-c'
 * removeAccentsAndSpecialCharacters('a b c', '_') // 'a_b_c'
 * removeAccentsAndSpecialCharacters('a b c', ' ') // 'a b c'
 * removeAccentsAndSpecialCharacters('     a   b         c   ', '.') // 'a.b.c'
 * removeAccentsAndSpecialCharacters('a,b,c') // 'a-b-c'
 * removeAccentsAndSpecialCharacters('a/b/c') // 'a-b-c'
 */
export function removeAccentsAndSpecialCharacters(text: string, replaceWith = '-') {
  const newText = remove(text)
    .normalize('NFKD')
    .replaceAll(/[^a-zA-Z0-9]/g, ' ')
  return splitTrimTruthy(newText, ' ').join(replaceWith)
}

/**
 * Splits `text` by `split`, trims each part, and removes any empty parts
 * @examples
 * splitTrimTruthy('a, b, c', ',') // ['a', 'b', 'c']
 * splitTrimTruthy('a, , c', ',') // ['a', 'c']
 * splitTrimTruthy('a: : c: : :: :', ':') // ['a', 'c']
 **/
export function splitTrimTruthy(text: string, split: string) {
  return text
    .split(split)
    .map((s) => s.trim())
    .filter((t) => t)
}
