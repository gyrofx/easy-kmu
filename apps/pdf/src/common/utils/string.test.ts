import { removeAccentsAndSpecialCharacters, splitTrimTruthy, summarizeStringTo } from './string'

describe('summarizeStringTo', () => {
  it.each([1, 10, 100, 199, 200, 201, 202, 203, 204, 300, 1000, 10000])(
    'summarizes a string of length %s to 200 characters',
    (length) => expect(summarizeStringTo(generateStringOfLength(length), 200)).toMatchSnapshot(),
  )

  it.each([1, 10, 100, 199, 200, 201, 202, 203, 204, 300, 1000, 10000])(
    'summarizes a string of length %s to 250 characters',
    (length) => expect(summarizeStringTo(generateStringOfLength(length), 250)).toMatchSnapshot(),
  )

  it.each([1, 10, 100, 199, 200, 201, 202, 203, 204, 300, 1000, 10000])(
    'throws an error if an invalid length is supplied with a string of length %s',
    (length) => expect(() => summarizeStringTo(generateStringOfLength(length), 150)).toThrow(),
  )
})

describe('splitTrimTruthy', () => {
  it.each([
    ['a, b, c', ',', ['a', 'b', 'c']],
    ['a, , c', ',', ['a', 'c']],
    ['a: : c: : :: :', ':', ['a', 'c']],
  ])('splits a string of %s by %s to %s', (text, split, expected) => {
    expect(splitTrimTruthy(text, split)).toEqual(expected)
  })
})

describe('removeAccentsAndSpecialCharacters', () => {
  it.each([
    ['abc', 'abc', '-'],
    ['a b c', 'a-b-c', '-'],
    ['a b c', 'a_b_c', '_'],
    ['a b c', 'a b c', ' '],
    ['     a   b         c   ', 'a.b.c', '.'],
    ['a,b,c', 'a-b-c', '-'],
    ['1234+20:00', '1234-20-00', '-'],
    ['äöü', 'aou', '-'],
    ['a, b, c', 'a-b-c', '-'],
    ['a/b/c', 'a-b-c', '-'],
  ])('removes accents from %s to %s', (text, expected, replaceWith) => {
    expect(removeAccentsAndSpecialCharacters(text, replaceWith)).toEqual(expected)
  })
})

function generateStringOfLength(length: number) {
  return someLongString.repeat(Math.ceil(length / someLongString.length) + 1).slice(0, length)
}

const someLongString =
  'abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3gw3hgw3gh9jawh30jgUOYNNUHBGYVTYFBHUNJWDIQJPMOIWFEHWEGWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$&^T&%*R&#(*Y@R(FYGHNEBWWGINUWEGIUH#(@G*NHUNWIEGFHIWEUGHFNWIEGUH(#*GHN(IUWESHGB*#&GHN(@GJOPIEGJ(*G$J@NPWOGJIWEMG)(#$GJM)W#GIJMWEG)(*#UGJN)J@#JGM)'
