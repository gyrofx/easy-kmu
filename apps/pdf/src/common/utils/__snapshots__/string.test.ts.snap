// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`summarizeStringTo > summarizes a string of length 1 to 200 characters 1`] = `"a"`;

exports[`summarizeStringTo > summarizes a string of length 1 to 250 characters 1`] = `"a"`;

exports[`summarizeStringTo > summarizes a string of length 10 to 200 characters 1`] = `"abcdefg123"`;

exports[`summarizeStringTo > summarizes a string of length 10 to 250 characters 1`] = `"abcdefg123"`;

exports[`summarizeStringTo > summarizes a string of length 100 to 200 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3g"`;

exports[`summarizeStringTo > summarizes a string of length 100 to 250 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3g"`;

exports[`summarizeStringTo > summarizes a string of length 199 to 200 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3gw3hgw3gh9jawh30jgUOYNNUHBGYVTYFBHUNJWDIQJPMOIWFEHWEGWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#"`;

exports[`summarizeStringTo > summarizes a string of length 199 to 250 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3gw3hgw3gh9jawh30jgUOYNNUHBGYVTYFBHUNJWDIQJPMOIWFEHWEGWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#"`;

exports[`summarizeStringTo > summarizes a string of length 200 to 200 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3gw3hgw3gh9jawh30jgUOYNNUHBGYVTYFBHUNJWDIQJPMOIWFEHWEGWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$"`;

exports[`summarizeStringTo > summarizes a string of length 200 to 250 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3gw3hgw3gh9jawh30jgUOYNNUHBGYVTYFBHUNJWDIQJPMOIWFEHWEGWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$"`;

exports[`summarizeStringTo > summarizes a string of length 201 to 200 characters 1`] = `"[Summarized string of length 201]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqc [...] GWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$&"`;

exports[`summarizeStringTo > summarizes a string of length 201 to 250 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3gw3hgw3gh9jawh30jgUOYNNUHBGYVTYFBHUNJWDIQJPMOIWFEHWEGWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$&"`;

exports[`summarizeStringTo > summarizes a string of length 202 to 200 characters 1`] = `"[Summarized string of length 202]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqc [...] WNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$&^"`;

exports[`summarizeStringTo > summarizes a string of length 202 to 250 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3gw3hgw3gh9jawh30jgUOYNNUHBGYVTYFBHUNJWDIQJPMOIWFEHWEGWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$&^"`;

exports[`summarizeStringTo > summarizes a string of length 203 to 200 characters 1`] = `"[Summarized string of length 203]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqc [...] NEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$&^T"`;

exports[`summarizeStringTo > summarizes a string of length 203 to 250 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3gw3hgw3gh9jawh30jgUOYNNUHBGYVTYFBHUNJWDIQJPMOIWFEHWEGWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$&^T"`;

exports[`summarizeStringTo > summarizes a string of length 204 to 200 characters 1`] = `"[Summarized string of length 204]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqc [...] EGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$&^T&"`;

exports[`summarizeStringTo > summarizes a string of length 204 to 250 characters 1`] = `"abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8465469846541a69g48w3ewh3gw3hgw3gh9jawh30jgUOYNNUHBGYVTYFBHUNJWDIQJPMOIWFEHWEGWNEGINOJPIEGWFHEWE*GUH*WG#UFEIJNWIOEGJN(#U*G(W#$&^T&"`;

exports[`summarizeStringTo > summarizes a string of length 300 to 200 characters 1`] = `"[Summarized string of length 300]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqc [...] WEUGHFNWIEGUH(#*GHN(IUWESHGB*#&GHN(@GJOPIEGJ(*G$J@"`;

exports[`summarizeStringTo > summarizes a string of length 300 to 250 characters 1`] = `"[Summarized string of length 300]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8 [...] INUWEGIUH#(@G*NHUNWIEGFHIWEUGHFNWIEGUH(#*GHN(IUWESHGB*#&GHN(@GJOPIEGJ(*G$J@"`;

exports[`summarizeStringTo > summarizes a string of length 1000 to 200 characters 1`] = `"[Summarized string of length 1000]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqc [...] H(#*GHN(IUWESHGB*#&GHN(@GJOPIEGJ(*G$J@NPWOGJIWEMG)"`;

exports[`summarizeStringTo > summarizes a string of length 1000 to 250 characters 1`] = `"[Summarized string of length 1000]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8 [...] G*NHUNWIEGFHIWEUGHFNWIEGUH(#*GHN(IUWESHGB*#&GHN(@GJOPIEGJ(*G$J@NPWOGJIWEMG)"`;

exports[`summarizeStringTo > summarizes a string of length 10000 to 200 characters 1`] = `"[Summarized string of length 10000]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqc [...] )W#GIJMWEG)(*#UGJN)J@#JGM)abcdefg1234567qn80hcfnq2"`;

exports[`summarizeStringTo > summarizes a string of length 10000 to 250 characters 1`] = `"[Summarized string of length 10000]: abcdefg1234567qn80hcfnq203urvqv27yr3bhfq8327ryvfqcru809q23utgn9aw8ugjaw3hg8 [...] J(*G$J@NPWOGJIWEMG)(#$GJM)W#GIJMWEG)(*#UGJN)J@#JGM)abcdefg1234567qn80hcfnq2"`;
